### ISSUE TITLE
Validation errors from presentation only appear if I create semaphores after creating the swapchain

### ISSUE THREAD

**Issue by McFlyboy** [KhronosGroup/Vulkan-ValidationLayers#4422]
**Describe the Issue**
I have set up a render loop using the Vulkan-Hpp raii headers with validation layers enabled. So far I'm only using a renderpass to clear the images before presentation. No pipelines or draw commands are being called.
As I'm just starting out I have been using `device.waitIdle()` for synchronization instead of using semaphores and fences.
However, the first two times I try to present an image by calling `queue.presentKHR()` I get a validation error stating that the image I'm trying present is in layout `VK_IMAGE_LAYOUT_UNDEFINED`. After receiving this error two times in a row everything suddenly works fine without error.

The result returned from `queue.presentKHR()` is `vk::Result::eSuccess` even when it fails. I have tried turning off the validation layers and stepping through the code frame for frame, and with the validation layers off the presentation works and the image shows every time, including the first two times.

However I have noticed that if I call `device.createSemaphore()` before creating the swapchain, even if I'm not using the semaphore for anything, then the error only show the first time instead of showing for the two first times like before. If I create a second semaphore, again without actually using it, then both errors are solved and everything runs smoothly. If I postpone the semaphore creation until after swapchain creation the errors come back again.

**Valid Usage ID**
`Validation Error: [ VUID-VkPresentInfoKHR-pImageIndices-01296 ] Object 0: handle = 0x24a81a31440, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0xc7aabc16 | vkQueuePresentKHR(): pSwapchains[0] images passed to present must be in layout VK_IMAGE_LAYOUT_PRESENT_SRC_KHR or VK_IMAGE_LAYOUT_SHARED_PRESENT_KHR but is in VK_IMAGE_LAYOUT_UNDEFINED. The Vulkan spec states: Each element of pImageIndices must be the index of a presentable image acquired from the swapchain specified by the corresponding element of the pSwapchains array, and the presented image subresource must be in the VK_IMAGE_LAYOUT_PRESENT_SRC_KHR layout at the time the operation is executed on a VkDevice (https://github.com/KhronosGroup/Vulkan-Docs/search?q=)VUID-VkPresentInfoKHR-pImageIndices-01296)`

**Environment:**
 - OS: Windows 11 Pro 21H2
 - GPU: NVIDIA GeForce GTX 1060 6GB
 - SDK or header version if building from repo: 1.3.216.0
 - Options enabled (synchronization, best practices, etc.): `VK_KHR_surface`, `VK_KHR_win32_surface`, `VK_EXT_debug_utils`

**Additional context**

**Comment by McFlyboy** [comment]
I tried out using semaphores and a fence instead of calling `device.waitIdle()`, and this worked just fine, until I decided to try using multiple frames in flight. I doubled the amount of semaphores and fences I create, and now I get the same problem again. This time I get the error only once if I create them before swapchain creation, and twice if I create them after swapchain creation.

Please have a look at this issue soon @jeremyg-lunarg, so that we can figure out if this is actually caused by an issue in the validation layers🙏

**Comment by jeremyg-lunarg** [comment]
Sorry for the delay.  This is most likely related to #4308. I have a candidate fix for that issue which has taken much more time than I thought to get working.  Are you able to try that fix? If not, could you provide some sample code, a gfxreconstruct capture or an api dump showing the error?

**Comment by McFlyboy** [comment]
Hi again. I haven't tried your fix yet cause I'm not sure how I can test it out. I did try out creating a gfxreconstruct capture though as you also suggested. Here's the capture: [gfxrecon_capture_20220922T215900.zip](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/9629085/gfxrecon_capture_20220922T215900.zip)

Hopefully it will be useful to figure out the issue. Let me know if you need anything else, and tell me if you find something! 😃

**Comment by jeremyg-lunarg** [comment]
This is a bug in the validation layers, not your code.  Image layouts are first tracked within Command Buffers, and then copied into 'global' state during vkQueueSubmit.  The global state for the swapchain image starts out as UNDEFINED. Your command buffer has a render pass which correctly sets the final layout  to VK_IMAGE_LAYOUT_PRESENT_SRC_KHR, as required by vkQueuePresentKHR. vkQueuePresentKHR uses the global state to check VUID 01296.  However until the first vkQueueSubmit call, the global state is UNDEFINED, after that it goes to VK_IMAGE_LAYOUT_PRESENT_SRC_KHR.   This is why you only get the error on the first frame.

I'm am not sure why changing the order in which you create semaphores and the swapchain, I would expect you to hit the bug every time.

The fix to validation to get rid of this bug is unfortunately very complicated.   All VUIDs that include a phrase such as "at the time the operation is executed on a VkDevice" must be checked when validation's state matches what the device would 'see' at execution time. I'm currently working on part of this problem in PR #4308, but more work will be required after that to move all of the execution time checks to happen at the correct time.

**Comment by McFlyboy** [comment]
Good to know that the problem isn't in my code. Now, what do we do from here? Should I provide you with a gfxreconstruct capture where I create the semaphores after swapchain creation as well? Also, I've noticed that when I re-create the swapchain after a window resize that I get the error either zero times, once or twice. Should I create a gfxreconstruct capture for that also, so that you have more info about the issue?

**Comment by jeremyg-lunarg** [comment]
I don't think I need any of that at this point, other than time to fix the problem. :sweat_smile:  The validation code for this is clearly wrong and knowing why it only fails sometimes won't help at this point. Having the first capture you sent was very helpful and eventually I'll make a unit test case based on it. For now you should just ignore this error, by putting it in the VK_MESSAGE_FILTER environment variable or vk_settings.txt or using vkconfig (see: https://vulkan.lunarg.com/doc/sdk/1.3.224.1/windows/khronos_validation_layer.html)

**Comment by kanthonye** [comment]
Temp


I had the same issue.

**pSwapchains[0] images passed to present must be in layout VK_IMAGE_LAYOUT_PRESENT_SRC_KHR or VK_IMAGE_LAYOUT_SHARED_PRESENT_KHR but is in VK_IMAGE_LAYOUT_UNDEFINED. The Vulkan spec states: Each element of pImageIndices must be the index of a presentable image acquired from the swapchain specified by the corresponding element of the pSwapchains array, and the presented image subresource must be in the VK_IMAGE_LAYOUT_PRESENT_SRC_KHR layout at the time the operation is executed on a VkDevice**

The issue was that I was creating a total of **MAX_FRAMES_IN_FLIGHT** frambuffers. My **MAX_FRAMES_IN_FLIGHT** is 2. this was my first mistake. My second mistake is using my **current_index** value which repeatedly increments from 0 to 1. How I found my mistakes is by debugging and stepping through my code. I realize that **_swapchain_imageviews.size = 3** while my **_framebuffers.size = 2**. The frame buffer size is supposed to be equal to the _swapchain_imageviews size. And the index for selecting the correct frame buffer is **image_index** which is acquired from **vkAcquireNextImageKHR**. Once I fixed the issue the error when away. 

**Here is my code for creating framebuffers before:**

```
_framebuffers.resize( MAX_FRAMES_IN_FLIGHT );
    for (int i=0; i<MAX_FRAMES_IN_FLIGHT; i++)
    {
      _framebuffers[i] = _core->create()->framebuffer( _extent, _swapchain_imageviews[i], _renderpass, _depth_imageview );
      if ( _framebuffers[i] == VK_NULL_HANDLE )
      {
        for (int k=0; k<i; k++)
        {
          _core->destroy()->framebuffer( _framebuffers[i] );
        }
        return false;
      }
    }
```

**Here is my code for creating framebuffers after the fix:**

```
_framebuffers.resize( _swapchain_imageviews.size() );
    for (int i=0; i<_swapchain_imageviews.size(); i++)
    {
      _framebuffers[i] = _core->create()->framebuffer( _extent, _swapchain_imageviews[i], _renderpass, _depth_imageview );
      if ( _framebuffers[i] == VK_NULL_HANDLE )
      {
        for (int k=0; k<i; k++)
        {
          _core->destroy()->framebuffer( _framebuffers[i] );
        }
        return false;
      }
    }
```


**Here is my code for VkRenderPassBeginInfo before:**


```
VkRenderPassBeginInfo renderPassInfo{};
renderPassInfo.sType = VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO;
renderPassInfo.renderPass = _renderpass;
renderPassInfo.framebuffer = _framebuffers[_current_frame];
renderPassInfo.renderArea.offset = {0, 0};
renderPassInfo.renderArea.extent = _extent;

std::array<VkClearValue, 2> clearValues{};
clearValues[0].color = {{0.02, 0.02, 0.0225, 1.0}};
clearValues[1].depthStencil = {1.0f, 0};

renderPassInfo.clearValueCount = static_cast<uint32_t>(clearValues.size());
renderPassInfo.pClearValues = clearValues.data();

vkCmdBeginRenderPass(cmd_bufr, &renderPassInfo, VK_SUBPASS_CONTENTS_INLINE);
```

**Here is my code for VkRenderPassBeginInfo after the fix:**

```
VkRenderPassBeginInfo renderPassInfo{};
renderPassInfo.sType = VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO;
renderPassInfo.renderPass = _renderpass;
renderPassInfo.framebuffer = _framebuffers[_image_index];
renderPassInfo.renderArea.offset = {0, 0};
renderPassInfo.renderArea.extent = _extent;

std::array<VkClearValue, 2> clearValues{};
clearValues[0].color = {{0.02, 0.02, 0.0225, 1.0}};
clearValues[1].depthStencil = {1.0f, 0};

renderPassInfo.clearValueCount = static_cast<uint32_t>(clearValues.size());
renderPassInfo.pClearValues = clearValues.data();

vkCmdBeginRenderPass(cmd_bufr, &renderPassInfo, VK_SUBPASS_CONTENTS_INLINE);
```

Anyway, I just wanted to leave this comment just Incase anyway ran into the same issue.

**Comment by McFlyboy** [comment]
@jeremyg-lunarg I have now discovered the bug in my code that is triggering this validation error!
I was using `std::map<vk::Image, vk::raii::ImageView>` in my code for pairing up images and image views, ignorant of the fact that an `std::map` orders its content based on the key (the memory address of the raw `VkImage`). This caused the framebuffers to end up being stored in the wrong order as well, which in turn led to the wrong framebuffer being picked for rendering and swapping at every frame!
I'm now using `std::vector<std::pair<vk::Image, vk::raii::ImageView>>` and it works perfectly! I have turned off the validation layer override that you taught me and I get no errors! 😃 

Is there still a bug in the validation layers as you said there was? Or should we just call it a user error and close the issue?

**Comment by Jens-Eckert** [comment]
Popping in to say I had the same issue and it was fixed by making MAX_FRAMES_IN_FLIGHT equal to the amount of swapchain images!!!

**Comment by artem-lunarg** [comment]
I think this is fixed now. After https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/10316 we account for submission order and also there were bunch of other image layouts fixes.

This error in the latest code will likely be related to swapchian synchronization which is not that hard to do, but it's easy to chose a wrong method. Still, if you read this comment and think it's VVL error for your use case, please open a new issue with a description a specific scenario and we will investigate.
